### PR TITLE
Add direct dependency on APEX

### LINKED ISSUE #40: Revive APEX support
With the distributed functionality removed, the actual APEX support was removed. APEX can be turned into a direct dependency with no special support required on the APEX side and is quite straightforward to implement. I have some old working code for this already but from HPX. This needs to be revived.

### PR DESCRIPTION
**Pull request by msimberg**
Fixes #40. Also see #103.

Still to do:
- [x] make sure CI has APEX installed and available
- [x] add patch to APEX spack package (see https://github.com/spack/spack/pull/28844, the patch is already applied on our local spack instance)

This also adds a way to use a spack environment in Jenkins, which is now used for the APEX build. Currently the only spack build is `gcc-10-apex-spack`. I've left the `-spack` suffix while this is the only configuration using spack. If we move all of them to use spack in the future we can remove the suffix.

I've disabled all other CI configurations that had APEX enabled (it was mostly formal, they didn't actually have real APEX support anyway). The only one with APEX enabled is now the new spack configuration on Jenkins.

## Proposed Changes

- Adds a direct dependency on APEX and uses its `task_wrapper` type directly. APEX does not depend on pika.
- Adds the `annotation_check` test from HPX.
- Does light refactoring of headers and functions in the `external_timer` namespace.
- Moves `external_timer` into `pika::detail`.

## Checklist

Not all points below apply to all pull requests.

- [x] I have added a new feature and have added tests to go along with it.
- [ ] I have fixed a bug and have added a regression test.
- [ ] I have added a test using random numbers; I have made sure it uses a seed, and that random numbers generated are valid inputs for the tests.

### CODE REVIEW

File: cmake/FindAPEX.cmake [hunk 1]
```diff
@@ -0,0 +1,72 @@
+# Copyright (c)      2014 Thomas Heller
+# Copyright (c) 2007-2012 Hartmut Kaiser
+# Copyright (c) 2010-2011 Matt Anderson
+# Copyright (c) 2011      Bryce Lelbach
+#
+# SPDX-License-Identifier: BSL-1.0
+# Distributed under the Boost Software License, Version 1.0. (See accompanying
+# file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+if(NOT TARGET pika_internal::apex)
```
Comment (msimberg) [inline cmake/FindAPEX.cmake:10]: Note that I'm not using the name `APEX::apex` here. I realized that we probably shouldn't be using namespaces without `pika` in our `FindX.cmake` files, and this is how I'd probably want to have our other `FindX.cmake` files changed as well. Only using `pika(_internal)` for things that we set up means that there's no chance of conflicts from other dependencies setting up targets within the same namespace. Low probability, but I think it's good hygiene.

File: cmake/pika_setup_apex.cmake [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2022 ETH Zurich
+#
+# SPDX-License-Identifier: BSL-1.0
+# Distributed under the Boost Software License, Version 1.0. (See accompanying
+# file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+if(PIKA_WITH_APEX)
+  pika_find_package(APEX REQUIRED)
```
Comment (msimberg) [inline cmake/pika_setup_apex.cmake:8]: This is also a change that would be nice to have for other dependencies as well.

File: cmake/FindAPEX.cmake (hunk 1 above)
Comment (msimberg) [reply cmake/FindAPEX.cmake:10]: It actually alread has an issue... #9.

**Comment by msimberg** [comment]
test cscs

**Comment by msimberg** [comment]
test cscs

**Comment by msimberg** [comment]
test cscs

**Comment by msimberg** [comment]
test cscs

**Comment by msimberg** [comment]
test cscs

**Comment by msimberg** [comment]
test cscs

**Comment by msimberg** [comment]
test cscs

**Comment by msimberg** [comment]
test cscs

**Comment by msimberg** [comment]
test cscs

**Comment by msimberg** [comment]
test cscs

**Comment by msimberg** [comment]
test cscs

**Comment by msimberg** [comment]
test cscs

**Comment by msimberg** [comment]
test cscs

**Comment by msimberg** [comment]
test cscs

File: .jenkins/cscs/batch.sh [hunk 3]
```diff
@@ -21,14 +21,38 @@ source "${src_dir}/.jenkins/cscs/env-common.sh"
...
-ctest \
-    --verbose \
-    -S ${src_dir}/.jenkins/cscs/ctest.cmake \
-    -DCTEST_BUILD_EXTRA_OPTIONS="${build_extra_options}" \
-    -DCTEST_CONFIGURE_EXTRA_OPTIONS="${configure_extra_options} -DCMAKE_INSTALL_PREFIX=${install_dir}" \
-    -DCTEST_BUILD_CONFIGURATION_NAME="${configuration_name_with_build_type}" \
-    -DCTEST_SOURCE_DIRECTORY="${src_dir}" \
-    -DCTEST_BINARY_DIRECTORY="${build_dir}"
+if [[ -n "${spack_spec:-}" ]]; then
+    # TODO: This is temporary
```
Comment (msimberg) [inline .jenkins/cscs/batch.sh:25]: Remove before merging.

**Comment by msimberg** [comment]
test cscs

**Comment by msimberg** [comment]
test cscs

**Review by aurianer — APPROVED** [review]
LGTM thanks!

**Comment by msimberg** [comment]
bors merge

**Pull request merged**